### DISCUSSION TITLE
Some UI.Handle and hand interaction thoughts

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#564] [Ideas]
I have a simple 3D model viewer that I demo quite often with an XR-3. It has a menu to switch between a few detailed models and this really works well to convey the power of XR, but something that people always try is to grab the models to inspect them. So far I have been holding off of enabling that type of interaction as it can get messy quickly (especially with the somewhat twitchy hand tracking on the XR-3). After experimenting a bit today I enabled grab interaction using `UI.Handle` but here's a few things I noticed:

* As part of the interaction I show the handle box as a line-based cube, to make people aware that grab is active, plus what the interaction volume is. I noticed that for many models the handle box is quite a bit larger than needed, looking at the model vertices. I think I'm drawing it correctly with:
```
                            UI.HandleBegin("ModelHandle", ref model_pose, scaled_bounds);
                            if (show_handle_box)
                                Mesh.Cube.Draw(Material.UIBox, Matrix.TS(scaled_bounds.center, scaled_bounds.dimensions));
```

* The previous point leads to somewhat tricky interaction as you can have your hand(s) within the handle volume quite easily and inadvertently make a grab gesture and transform the model. To overcome this I use a button on the left-hand palm to enable grab mode (disabled by default), plus disable the `UI.HandleBegin()` path when that button is visible, so you can always get back to a state where the model stays in place and you can disable grab. But what would help as well would be a way to limit handle interaction to a single hand, e.g only right. As moving your left hand to get to the grab palm menu can already cause grab interaction on the XR-3 apparently (and the hand tracking volume extents quite a bit further down below the FOV than I expected). The handle interaction also sometimes seems to keep some state related to the left hand when interacting with the right hand, as the rotation pivot should be the initial grab position, right? The thing I notice sometimes is the pivot being close to the left hand position while I'm grabbing with the right hand.

* A way to signal to signal interaction is about to happen could be to give the hands a different color (e.g. turn to red). I think I should be able to change the hand material for hand-is-about-to-grab but don't see an easy way to reliably test for this. As that involves testing several of the hand features I suspect (and not just palm position)?

**Comment by maluoi** [comment]
I've been thinking about this a bunch recently as well! I do think there's need for some extra criteria around Handles and interaction here, I've been seeing a lot of false positives while using Quest Pro.

> I noticed that for many models the handle box is quite a bit larger than needed, looking at the model vertices.

The bounding box is probably correct? Bounding boxes are axis aligned, which have a tendency to have a lot of extra volume to them. Stray invisible/small geo could also throw the bounds as well. I believe SK does also expand the Handle volume by a centimeter or so to help with grab expectations? I've been re-evaluating that, but it's probably not what you're seeing here.

> The previous point leads to somewhat tricky interaction as you can have your hand(s) within the handle volume quite easily and inadvertently make a grab gesture and transform the model.

I definitely also have used an "edit mode" vs a "locked mode" to help with this. You _could_ be interested in defining or calculating more conservative bounds for some `Model`s? Or you might also add grab handles to the asset rather than making the whole asset grabbable. This could be an occasion where a small centrally located transform gizmo/widget might be good, or a more UI encrusted bounding box widget.

The UI Box used to have a ring show up when the finger went through it, which would help, but I think I broke that at some point and haven't had a chance to add it back!

> As moving your left hand to get to the grab palm menu can already cause grab interaction on the XR-3 apparently (and the hand tracking volume extents quite a bit further down below the FOV than I expected).

I've been thinking of adding some FoV checks for hand interaction? Like pinch actions for hands below a certain angle might get discarded, requiring hands to be up and "intentful". That's probably a bit of a fuzzy idea, so I'm still trying to decide what else needs to go into that. I'll probably be discussing this with some of my MRTK co-workers next week, I believe they may have a bit of experience here. I'd also be interested if you have suggestion here too :)

Explicitly disallowing 2H interactions could be a good addition too, but I'd love for it to to be unnecessary in a situation like this? Hopefully the rest of the criteria additions can be enough!

> A way to signal to signal interaction is about to happen could be to give the hands a different color (e.g. turn to red). I think I should be able to change the hand material for hand-is-about-to-grab but don't see an easy way to reliably test for this. As that involves testing several of the hand features I suspect (and not just palm position)?

Right now, I think the best tool for this may be `UI.LastElementFocused`? Which might be a bit too surgical in this case, and doesn't exactly indicate which hand. I peeked at Cubism's far interact action recently, and kinda fell in love with the way they glow the fingertips there! I might investigate doing more stuff like this :)

I do need to find a way to do this that still allows the user control over the hand's visual state though, I think there's some potential conflict here that may require slightly larger architecture.

**Comment by laultman** [comment]
I too have these concerns using the HoloLens 2. The inadvertent movement of larger objects where the user may actually be "inside" bounds. For example, I have a use case where the user if fitting a 1:1 scale 8 cylinder turbo diesel engine in its engine bay for fit and alignment of other vehicle systems. To do precision movements aligning engine mount holes to real-world frame holes is a difficult task because you are too close or inside the object's bounds. Any twitch moves the engine. I need a snap/lock tool :)
